File: OCaml/Benchmark/perfect_number.ml
(* vim: set ft=ocaml sw=2 ts=2: *)

(* perfect_number.ml
 * A package to calculate Perfect numbers
 *)

(* predicate to check for a perfect number *)
let is_perfect n =
  let rec loop i sum =
    if n = i then sum = n else
      if n mod i = 0 then loop (i+1) (sum+i) else loop (i+1) sum
  in
  loop 1 0


(* generate a list of perfect numbers until given upper limit *)
let perfect_numbers n =
  let rec loop i =
    if i = n then [] else
      if is_perfect i then i :: loop (i+1) else loop (i+1)
  in
  loop 1

File: Julia/Benchmark/src/PerfectNumber.jl
# vim: set ft=julia sw=4 ts=4:
#
# PerfectNumber.jl
#
# Calculate 'perfect' numbers in Julia

module PerfectNumber

export isPerfect


"Predicate to check whether 'n' is perfect"
function isPerfect(n)
	sum = 0
	for i = 1:(n-1)
		if mod(n, i) == 0
			sum += i
		end
	end
	sum == n
end



end # module


